### PR TITLE
	Remove JoinableTaskFactory.Run usage in GetRule

### PR DESCRIPTION
**Pull request by davkean**
**Customer scenario**

After installing one of the 15.3 Previews customers will start experiencing more "random" UI delays in normal VS operations including:

- Expanding nodes in Solution Explorer
- Add new and removing items to and from the tree
- Opening C# source files
- Opening XML files
- Typing
- Opening and closing the solution

**Bugs this fixes:** 

Part of the general effort of chasing down the UI delays introduced in the project system in 15.3: https://github.com/dotnet/project-system/issues/2297.

This particular case is tracked by: https://github.com/dotnet/project-system/issues/2389 and affects cross-targeting projects..

**Workarounds, if any**

None

**Risk**

Low. This simply replaces an _async -> synchronous -> async_ call with _async -> async_.

**Performance impact**

This will reduce UI delays and reduce the number of active threads (we were blocking ~20 threadpool threads due to the call)

**Is this a regression from a previous update?**

Yes, the dependency node was rewritten in 15.3.

**Root cause analysis:**

Threading is hard. Threading in VS is even harder. Here we were doing a blocking call from a ThreadPool thread that was blocked on a lock that the UI thread held, which itself was blocked on thread pool creation. This issue is discussed at length here: https://github.com/Microsoft/vs-threading/blob/v15.3/doc/threadpool_starvation.md.

We're still learning how this all works, but we're [talking about strategies (compilation errors)](https://github.com/dotnet/project-system/issues/2391) to prevent these sorts of issues in the future. They are hard to find, diagnose and write tests for.

**How was the bug found?**

Many internal reports both from ASP.NET, Roslyn, ProjectSystem itself and opening external projects (such as https://github.com/aarnott/pinvoke).

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs [hunk 1]
```diff
@@ -390,17 +390,18 @@ private Task BuildTreeForSnapshotAsync(IDependenciesSnapshot snapshot)
             }
 
             var nowait = SubmitTreeUpdateAsync(
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs]: @lifengl I noticed this existing "nowait" pattern here - should we be Forking/FireAndForgeting this thing?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs [hunk 2]
```diff
@@ -574,11 +575,8 @@ public IRule GetRule(IDependency dependency, IProjectCatalogSnapshot catalogs)
...
             else
             {
-                ThreadHelper.JoinableTaskFactory.Run(async () =>
-                {
-                    project = await DependenciesHost.GetConfiguredProject(dependency.TargetFramework)
-                                                    .ConfigureAwait(false) ?? ActiveConfiguredProject;
-                });
+                project = await DependenciesHost.GetConfiguredProject(dependency.TargetFramework)
+                                                .ConfigureAwait(false) ?? ActiveConfiguredProject;
             }
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs]: Actual code change is here, rest is just making it async all the way up.

**Comment by davkean** [comment]
tag @dotnet/project-system

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/TreeView/GroupedByTargetTreeViewProviderTests.cs [hunk 3]
```diff
@@ -32,7 +33,7 @@ public void GrouppedByTargetTreeViewProvider_WhenEmptySnapshot_ShouldJustUpdateD
 
             // Act
             var provider = new GroupedByTargetTreeViewProvider(treeServices, treeViewModelFactory, commonServices);
-            var resultTree = provider.BuildTree(dependenciesRoot, snapshot);
+            var resultTree = await provider.BuildTreeAsync(dependenciesRoot, snapshot);
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/TreeView/GroupedByTargetTreeViewProviderTests.cs]: ConfigureAwaits for these? Once we reenable analyzers it'll complain about these.

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/TreeView/GroupedByTargetTreeViewProviderTests.cs]: These are tests...

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/TreeView/GroupedByTargetTreeViewProviderTests.cs]: I did it elsewhere.

Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/TreeView/GroupedByTargetTreeViewProviderTests.cs]: I thought we ran the analyzer for tests too but actually [we don't](https://github.com/dotnet/project-system/blob/master/build/Rulesets/NonShippingProjectRules.ruleset#L8) - so ignore the comment.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Models/DependenciesViewModelFactory.cs [hunk 4]
```diff
@@ -32,8 +32,9 @@ public IDependencyViewModel CreateTargetViewModel(ITargetedDependenciesSnapshot
 
         public IDependencyViewModel CreateRootViewModel(string providerType, bool hasUnresolvedDependency)
         {
-            var provider = GetProvider(providerType);
-            return provider.CreateRootDependencyNode().ToViewModel(hasUnresolvedDependency);
+            IProjectDependenciesSubTreeProvider provider = GetProvider(providerType);
+            IDependencyModel dependencyModel = provider.CreateRootDependencyNode();
+            return dependencyModel.ToViewModel(hasUnresolvedDependency);
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Models/DependenciesViewModelFactory.cs]: Why this change?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs (hunk 1 above)
Comment (lifengl) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs]: SubmitTreeUpdateAsync pushes the change to the queue, and those requests will be processed in the future.  Most tree provider doesn't wait on that call.  (Unless you want to wait to make sure the tree has been updated, you don't need wait).  Those requests will be processed one at a time based on the order.

**Comment by lifengl** [comment]
changes look good to me.

**Comment by srivatsn** [comment]
Tagging @MattGertz for approval.

**Comment by MattGertz** [comment]
@davkean -- JoC has approved, but would like to understand the risk better.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Models/DependenciesViewModelFactory.cs (hunk 4 above)
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Models/DependenciesViewModelFactory.cs]: Because I set breakpoints debugging this - couldn't set conditional breakpoints, see the commit: https://github.com/dotnet/project-system/pull/2392/commits/fcdeac4a1c5a1b9e2208c4cbf6d5066a02eb1168.
